remote-config/src/fetch/fetcher.rs

```diff
@@ -0,0 +1,826 @@
+        // (target_files.get()) and the insertion later on. Makes more sense to just hold it
+        // continuously
+        let mut target_files = self.state.target_files_by_path.lock().unwrap();
+
+        let mut config_paths: HashSet<RemoteConfigPathRef<'static>> = HashSet::new();
+        for path in response.client_configs.iter() {
+            match RemoteConfigPath::try_parse(path) {
+                // SAFTEY: The lifetime of RemoteConfigPathRef is tied to the config_paths
+                // Vec<String>
+                Ok(parsed) => {
+                    config_paths.insert(unsafe {
+                        transmute::<RemoteConfigPathRef<'_>, RemoteConfigPathRef<'_>>(parsed)
+                    });
+                }
+                Err(e) => warn!("Failed parsing remote config path: {path} - {e:?}"),
+            }
+        }
+
+        if self.state.expire_unused_files {
+            target_files.retain(|k, _| config_paths.contains(&(&**k).into()));
+        }
+
+        for (path, target_file) in targets_list.signed.targets {
+            fn hash_sha256(v: &[u8]) -> String {
+                format!("{:x}", Sha256::digest(v))
+            }
+            fn hash_sha512(v: &[u8]) -> String {
+                format!("{:x}", Sha512::digest(v))
+            }
+            let (hasher, hash) = if let Some(sha256) = target_file.hashes.get("sha256") {
+                (hash_sha256 as fn(&[u8]) -> String, *sha256)
+            } else if let Some(sha512) = target_file.hashes.get("sha512") {
+                (hash_sha512 as fn(&[u8]) -> String, *sha512)
+            } else {
+                warn!("Found a target file without hashes at path {path}");
+                continue;
+            };
+            let parsed_path = match RemoteConfigPath::try_parse(path) {
+                Ok(parsed_path) => parsed_path,
+                Err(e) => {
+                    warn!("Failed parsing remote config path: {path} - {e:?}");
+                    continue;
+                }
+            };
+            let handle = if let Some(StoredTargetFile {
+                hash: old_hash,
+                handle,
+                ..
+            }) = target_files.get(&parsed_path as &dyn RemoteConfigPathType)
+            {
+                if old_hash == hash {
+                    continue;
+                }
+                Some(handle.clone())
+            } else {
+                None
+            };
+            if let Some(raw_file) = incoming_files.get(path) {
+                if let Ok(decoded) = base64::engine::general_purpose::STANDARD.decode(raw_file) {
+                    let computed_hash = hasher(decoded.as_slice());
+                    if hash != computed_hash {
+                        warn!("Computed hash of file {computed_hash} did not match remote config targets file hash {hash} for path {path}: file: {}", String::from_utf8_lossy(decoded.as_slice()));
+                        continue;
```

Comment:
I can make the whole thing an error too, sure.

Reply:
I'm looking over the system tests and I actually don't think have a test for this 👀. 

In theory this should never happen unless somebody is manipulating data in the middle. If it was happening due to an RC bug we have a catastrophic error in our system. (it's likely to be caught well before it gets to the tracer)

Since we don't have a system test, we can't really hold tracers to any implementation at this time. I think I should add one and we should get convergence from tracers - my recommendation is to follow the RFC and the behavior we will eventually enforce via tests, and if one hash is bad fail the entire update so that this gets fast reported up the chain. (we have logs for client failures and monitoring for major issues like this)

Reply:
Promoted to error.